### ISSUE TITLE
More advanced log configuration

### ISSUE THREAD

**Issue by ThomasArdal** [Azure/azure-functions-host#4405]
`Microsoft.Azure.Functions.Extensions` brought us the first attempt of being able to configure custom logging providers in Functions. With the current solution (where I can inject an `ILoggerProvider`), I can configure functions to log everything to a single log output. I would like to be able to do much more advanced logging configuration, though. Being able to register multiple logging provider, decide which log levels goes into which logs, and other features like that, is an essential part of most applications today.

Ideally, logging would be configured like any other application using Microsoft.Extensions.Logging:

```csharp
[assembly: FunctionsStartup(typeof(MyNamespace.Startup))]

namespace MyNamespace
{
    public class Startup : FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            ...
            builder.Services.AddLogging(logging =>
            {
                logging.AddConfiguration(config.GetSection("Logging"));
                logging.AddConsole();
                logging.AddDebug();
                logging.AddEventSourceLogger();
            });
        }
    }
}
```

**Comment by alrod** [comment]
We do not have plans to support custom logging providers in functions currently. @fabiocav, correct me if I am not right.

**Comment by fabiocav** [comment]
That is actually supported with the new DI capabilities.
The documentation here shows the explicit registration of a logger provider: https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-dependency-injection

We're working on more detailed documentation for this scenario

**Comment by ThomasArdal** [comment]
@fabiocav It's perfectly fine to close this issue if it isn't something you are planning to work on. But for future views on this issue, I just want to point out that logging as described in the original question isn't supported with the current DI capabilities. Unless I've misunderstood something, you can only register a logger provider but not actually log to multiple locations, configure which log levels goes where, and all of the other features of Microsoft.Extensions.Logging.

**Comment by fabiocav** [comment]
@ThomasArdal that's fair. A lot of what you've mentioned above is supported and we're lacking documentation on how to do that with what is currently exposed. Can you clarify your distinction between registering a logger provider and logging to multiple locations?

I'll reopen that as requested so we can track the documentation updates and any other potential work.

**Comment by ThomasArdal** [comment]
@fabiocav Great. I'll look out for the documentation. I was looking to do something like this, which either isnt't supported or I cannot get to work:

```csharp
builder.Services.AddLogging(logging =>
{
    logging.AddConfiguration(config.GetSection("Logging"));
    logging.AddConsole();
    logging.AddDebug();
    logging.AddEventSourceLogger();
});
```

**Comment by kemmis** [comment]
Same here... can anyone show how to accomplish what @ThomasArdal is trying to do? Currently my logging is only going into app insights, but doesn't show up in the azure portal log window. I only see the built-in messages.

**Comment by ryanspletzer** [comment]
Not to pile up on this issue, but I know that a function will accept an `ILogger` in its method params and Azure Functions will provide it, but if another injected service wants an `ILogger` what is the best way to get that instance to it?

I am fine with the app insights-based `ILogger` instance being provided to the function method params to be provided elsewhere, but it seems that if I create and register a service with a dependency on `ILogger` it doesn't get that same instance.

**Comment by ryanspletzer** [comment]
Sorry I just noticed someone had posted about this over on the extensions repo where the DI extension lives: https://github.com/Azure/azure-functions-dotnet-extensions/issues/10

**Comment by ThomasArdal** [comment]
Any news on an example using custom logging from Azure Functions?

**Comment by ThomasArdal** [comment]
@fabiocav This issue can be closed (IMO). I got this working by [changing](https://github.com/elmahio/Elmah.Io.Extensions.Logging/commit/204a56f6f8d7b85c1fd71e4e3e038dd1a974ae81#diff-4ad396d41ea809f48e3d92b93648febf) the way my `ILoggerProvider` is registered.

For anyone else interested in seeing how to configure a custom `ILoggerProvider` with Azure Functions, there's a sample available here: https://github.com/elmahio/Elmah.Io.Extensions.Logging/tree/master/samples/Elmah.Io.Extensions.Logging.Function.

**Comment by kemmis** [comment]
One thought - if we haven't added more detailed examples in the project / documentation on MS docs, then perhaps we can take several of the ones listed in this issue (as well as the IConfiguration issue I started), and put them in the project where most devs go to look?

**Comment by ThomasArdal** [comment]
Some extended documentation would be nice. Also, if there are some additional requirements for how to implement a new `ILoggerProvider` in order for it to work with Functions, it would be nice to write them somewhere. I don't know if my provider was just implemented poorly, but it worked with both console and ASP.NET Core in the old version. So, something definitely works differently there when including it in Functions.

**Comment by kemmis** [comment]
Maybe someone around here can point us in the right direction of how to make this happen. Can we get one or two new issues created and label them as required documentation tasks. Then there we can discuss proper examples, and eventually get pull requests made out of them.

**Comment by ThomasArdal** [comment]
Good idea. I think this issue should be closed once the new issues have been created. This issue signals that custom providers aren't supported, which is not the case 😄

**Comment by fabiocav** [comment]
Thank you, all!

Will leave this open and flag this for follow up with @jeffhollan so we can prioritize documentation.
